src/openforms/utils/json_logic/descriptions.py

```diff
@@ -0,0 +1,325 @@
+        )
+        return f"[{joined_descriptions}]"
+
+    if isinstance(tree, Primitive):
+        if isinstance(tree, bool):
+            return repr(tree).lower()
+        return repr(tree)
+
+    # check for a baked in description
+    if meta_description := glom(
+        tree.source_expression, "_meta.description", default=""
+    ):
+        return str(meta_description)
+
+    _description_generator = OPERATION_DESCRIPTION_BUILDERS[tree.operator]
+    return _description_generator(tree, for_root=root)
+
+
+def _fmt_var(var: str) -> str:
+    return f"{{{{{var}}}}}"
+
+
+class JoinArgs:
+    def __init__(self, separator: str):
+        self.separator = separator
+
+    def __call__(self, operation: Operation, for_root: bool = False) -> str:
+        nested = [_generate_description(argument) for argument in operation.arguments]
+        description = self.separator.join(nested)
+        if not for_root:
+            description = f"({description})"
+        return description
+
+
+class TemplatedArgs:
+    def __init__(self, template: "_StrPromise", arg_names: Sequence[str]):
+        self.template = template
+        self.arg_names = arg_names
+
+    def _get_template_format_kwargs(self, args):
+        assert (num_args := len(args)) == (
+            num_names := len(self.arg_names)
+        ), f"Unexpected number of operation arguments, got {num_args}, expected {num_names}"
+        return {name: value for name, value in zip(self.arg_names, args)}
+
+    def __call__(self, operation: Operation, for_root: bool = False) -> str:
+        args = [_generate_description(argument) for argument in operation.arguments]
+        format_kwargs = self._get_template_format_kwargs(args)
+        description = self.template.format(**format_kwargs)
+        if not for_root:
+            description = f"({description})"
+        return description
+
+
+class FunctionLike(TemplatedArgs):
+    def __init__(self, template: "_StrPromise"):
+        self.template = template
+        super().__init__(template, arg_names=[])
+
+    def _get_template_format_kwargs(self, args):
+        return {"args": ", ".join(args)}
+
+    def __call__(self, operation: Operation, for_root: bool = False) -> str:
+        return super().__call__(operation, for_root=True)
+
+
+def add_boilerplate(wrap: bool = True, nested_as_root: bool = False):
+    def decorator(func) -> DescriptionGeneratorProtocol:
+        @wraps(func)
+        def _get_description(operation: Operation, for_root: bool = False):
+            args = [
+                _generate_description(argument, root=nested_as_root)
+                for argument in operation.arguments
+            ]
+            description = func(*args)
+            if not for_root and wrap:
+                description = f"({description})"
+            return description
+
+        return _get_description
+
+    return decorator
+
+
+def op_var(operation: Operation, for_root=False) -> str:
```

Comment:
Id the option `for_root` used anywhere?

Reply:
not used here, but the signature must match the [protocol](https://github.com/open-formulieren/open-forms/blob/818c34b0e5a038c8f7c5457a0e8c4a9e298acd7c/src/openforms/utils/json_logic/descriptions.py#L18) - other callbacks do require it.